### PR TITLE
refactor(sc-executor): use wasm executor builder instead of old apis

### PR DESCRIPTION
**Pull request by yjhmelody**
- Use the `builder` style apis to create runtime executor
- Deprecated the old `new` api for executor
- Improve and fix inconsistency about executor config(e.g. confused `max_runtime_instances`  with `runtime_cache_size`).
- Add `new_native_executor` to reduce boilerplate code.

--- 
polkadot address: 15ouFh2SHpGbHtDPsJ6cXQfes9Cx1gEFnJJsJVqPGzBSTudr

### CODE REVIEW

File: client/executor/src/executor.rs [hunk 1]
```diff
@@ -107,7 +105,7 @@ impl<H> WasmExecutorBuilder<H> {
 			onchain_heap_alloc_strategy: None,
 			offchain_heap_alloc_strategy: None,
 			max_runtime_instances: 2,
-			runtime_cache_size: 4,
+			runtime_cache_size: 8,
```
Comment (bkchr) [inline client/executor/src/executor.rs:108]: Please revert

File: client/executor/src/executor.rs [hunk 2]
```diff
@@ -238,6 +236,8 @@ impl<H> WasmExecutor<H>
 where
 	H: HostFunctions,
 {
+	/// *To be deprecated*: use [`Self::builder`] method instead of it.
+	///
```
Comment (bkchr) [inline client/executor/src/executor.rs:240]: Use a proper `deprecate` attribute

File: client/executor/src/executor.rs [hunk 3]
```diff
@@ -579,6 +581,8 @@ impl<D: NativeExecutionDispatch> NativeElseWasmExecutor<D> {
 		Self { native_version: D::native_version(), wasm: executor }
 	}
 
+	/// *To be deprecated*: use [`Self::new_with_wasm_executor`] method to configure it.
+	///
```
Comment (bkchr) [inline client/executor/src/executor.rs:585]: Same

File: client/executor/src/executor.rs [hunk 4]
```diff
@@ -539,6 +539,8 @@ pub struct NativeElseWasmExecutor<D: NativeExecutionDispatch> {
 }
 
 impl<D: NativeExecutionDispatch> NativeElseWasmExecutor<D> {
+	/// *To be deprecated*: use [`Self::new_with_wasm_executor`] method instead of it.
```
Comment (bkchr) [inline client/executor/src/executor.rs:542]: Same

**Review by bkchr — COMMENTED** [review]
Ty!

File: client/executor/src/executor.rs (hunk 1 above)
Comment (yjhmelody) [reply client/executor/src/executor.rs:108]: I want to know why? 
I see the `8` value everywhere, but the new api set `4` as default value.

Comment (yjhmelody) [reply client/executor/src/executor.rs:108]: I want to know why? 
I see the `8` value everywhere, but the new api set `4` as default value.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2612662

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2612661

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2618236

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2618235

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2619962

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2619963

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2620212

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2621123

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2621398

**Comment by yjhmelody** [comment]
@bkchr  Hi, anything need to be improved?

**Review by koute — COMMENTED** [review]
Looks mostly fine to me.

File: client/service/src/builder.rs [hunk 5]
```diff
@@ -228,6 +230,27 @@ where
...
+
+/// Creates a [`WasmExecutor`] according to [`Configuration`].
+pub fn new_wasm_executor<H: HostFunctions>(config: &Configuration) -> WasmExecutor<H> {
+	let strategy = config
+		.default_heap_pages
+		.map_or(DEFAULT_HEAP_ALLOC_STRATEGY, |p| HeapAllocStrategy::Static { extra_pages: p as _ });
+	WasmExecutor::<H>::builder()
+		.with_execution_method(config.wasm_method)
+		.with_onchain_heap_alloc_strategy(strategy)
+		.with_onchain_heap_alloc_strategy(strategy)
```
[the remark this answers is not in the document]
Comment (yjhmelody) [reply client/service/src/builder.rs:248]: Should be offchain

File: client/service/src/builder.rs [hunk 6]
```diff
@@ -228,6 +230,27 @@ where
 	Ok((client, backend, keystore_container, task_manager))
 }
 
+/// Creates a [`NativeElseWasmExecutor`] according to [`Configuration`].
+pub fn new_native_executor<D: NativeExecutionDispatch>(
```
[the remark this answers is not in the document]
Comment (yjhmelody) [reply client/service/src/builder.rs:234]: Ok, though I find somewhere use this name when use `NativeElseWasmExecutor`.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2650740

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2650741

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2650791

File: client/service/src/builder.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (koute) [reply client/service/src/builder.rs]: Yeah, our current naming might not be the best. (But we're going to remove the native executor soon-ish, so I guess it doesn't really matter all that much.)

**Comment by yjhmelody** [comment]
@koute Any other problem?

**Review by koute — APPROVED** [review]

**Review by bkchr — APPROVED** [review]

**Comment by bkchr** [comment]
bot merge

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/april-updates-for-substrate-and-polkadot-devs/2764/1

**Pull request merged**